src/google/protobuf/descriptor.cc

```diff
@@ -5480,6 +5489,75 @@ void DescriptorBuilder::BuildMessage(const DescriptorProto& proto,
   }
 }
 
+std::string GetJSONName(const FieldDescriptorProto& field, bool use_custom, bool* was_custom) {
+  if (use_custom && field.has_json_name()) {
+    *was_custom = true;
+    return field.json_name();
+  }
+  *was_custom = false;
+  return ToJsonName(field.name());
+}
+
+void DescriptorBuilder::CheckFieldJSONNameUniqueness(
+    const DescriptorProto& proto, const Descriptor* result) {
+  bool is_proto2 = result->file()->syntax() == FileDescriptor::SYNTAX_PROTO2;
+  std::string message_name = result->full_name();
+  // two passes: one looking only at default JSON names, and one that considers custom JSON names
+  CheckFieldJSONNameUniqueness(message_name, proto, is_proto2, false);
+  CheckFieldJSONNameUniqueness(message_name, proto, is_proto2, true);
+}
+
+struct JSONNameDetails {
+  const FieldDescriptorProto* field;
+  std::string orig_name;
+  bool is_custom;
+};
+
+void DescriptorBuilder::CheckFieldJSONNameUniqueness(
+    std::string message_name,const DescriptorProto& message, bool is_proto2, bool use_custom_names) {
+
+  std::map<std::string, struct JSONNameDetails> name_to_field;
+  for (int i = 0; i < message.field_size(); ++i) {
+    bool is_custom;
+    std::string name = GetJSONName(message.field(i), use_custom_names, &is_custom);
+    std::string lowercase_name = absl::AsciiStrToLower(name);
+    auto existing = name_to_field.find(lowercase_name);
+    if (existing != name_to_field.end()) {
+      auto match = existing->second;
```

Comment:
Can you give this an explicit type?

Reply:
Yep, sorry. I had used `auto` because at one point I had formulated this where it was the iterator (which has umpteen template parameters...). But the type is simple and clear now, so I'll change.

Reply:
done